docs: add better-auth guide for authentication (#9703)
withastro/docs: src/content/docs/en/guides/authentication.mdx

- [`auth-astro` on GitHub](https://github.com/nowaythatworked/auth-astro?tab=readme-ov-file#auth-astro)
- [Auth.js documentation](https://authjs.dev/)

## Better Auth

> Better Auth is framework-agnostic authentication (and authorization) framework for TypeScript. It provides a comprehensive set of features out of the box and includes a plugin ecosystem that simplifies adding advanced functionalities.

Reviewer: nit: not a native speaker but I would add:
```suggestion
Better Auth is a framework-agnostic authentication (and authorization) framework for TypeScript. It provides a comprehensive set of features out of the box and includes a plugin ecosystem that simplifies adding advanced functionalities.
```

> It supports astro out of the box, and you can use it to add authentication to your astro project.

Reviewer: ```suggestion
It supports Astro out of the box, and you can use it to add authentication to your Astro project.
```

### Installation

> ```shell
> npm install better-auth
> ```

Reviewer: ````suggestion
<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm install better-auth
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm add better-auth
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn add better-auth
  ```
  </Fragment>
</PackageManagerTabs>
````

For detailed setup instructions, check out the [Installtion Guide](https://www.better-auth.com/docs/installation).

### Configuration

Once you have installed and configured Better Auth in your project, you'll need to mount Better Auth handler in your Astro project.

```ts title="page/api/auth/[...all].ts"
import { auth } from "~/auth"; // import your Better Auth instance
import type { APIRoute } from "astro";
 
export const ALL: APIRoute = async (ctx) => {
	return auth.handler(ctx.request);
};
```

> Follow [Better Auth Astro Guide](https://www.better-auth.com/docs/integrations/astro) to learn more.

Reviewer: ```suggestion
Follow the [Better Auth Astro Guide](https://www.better-auth.com/docs/integrations/astro) to learn more.
```

- [`auth-astro` on GitHub](https://github.com/nowaythatworked/auth-astro?tab=readme-ov-file#auth-astro)
- [Auth.js documentation](https://authjs.dev/)

## Better Auth

Better Auth is a framework-agnostic authentication (and authorization) framework for TypeScript. It provides a comprehensive set of features out of the box and includes a plugin ecosystem that simplifies adding advanced functionalities.

It supports Astro out of the box, and you can use it to add authentication to your astro project.

### Installation

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm install better-auth
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm add better-auth
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn add better-auth
  ```
  </Fragment>
</PackageManagerTabs>

For detailed setup instructions, check out the [Installtion Guide](https://www.better-auth.com/docs/installation).

### Configuration

Once you have installed and configured Better Auth in your project, you'll need to mount Better Auth handler in your Astro project.

```ts title="page/api/auth/[...all].ts"

> import { auth } from "~/auth"; // import your Better Auth instance

Reviewer 2: ```suggestion
import { auth } from "../......"; // import your Better Auth instance
```
Can we make this a relative path instead of an import alias? copying this code won't work if people don't have that import configured.

This might mean, for the purposes of this example, we need a step where this is actually shown, creating a specific file to do so.

import type { APIRoute } from "astro";
 
export const ALL: APIRoute = async (ctx) => {
	return auth.handler(ctx.request);
};
```

Follow the [Better Auth Astro Guide](https://www.better-auth.com/docs/integrations/astro) to learn more.

### Usage

Better Auth offers clients for various frameworks, including Vanilla JS, React, Vue, Svelte, and Solid. To create a client for your chosen framework:

```ts title="src/lib/auth-client.ts"
import { createAuthClient } from 'better-auth/client';

export const authClient = createAuthClient({

>   // your configuration

Author: That wasn't really necessary. The configuration part comes once they are advanced with the library. Specifically to add client plugins. Which they should refer from Better Auth docs instead.

});

export const { signIn, signOut } = authClient;
```

> Once your client is set up, you can use it to authenticate users in your Astro components or any framework-specific files.

Reviewer 2: ```suggestion
Once your client is set up, you can use it to authenticate users in your Astro components or any framework-specific files. The following example adds the ability to log in or log out with your configured `signIn()` and `signOut()` functions.
```
Again, make all these little descriptions accurate, but where possible, tiny statements like this are helpful to guide the reader to what they should be looking for.

```astro title="src/pages/index.astro"
---
import Layout from 'src/layouts/Base.astro';
---
<Layout>
  <button id="login">Login</button>
  <button id="logout">Logout</button>

  <script>
    const { signIn, signOut } = await import("./lib/auth-client")
    document.querySelector("#login").onclick = () => signIn.social({
      provider: "github",
      callbackURL: "/dashboard",
    })
    document.querySelector("#logout").onclick = () => signOut()
  </script>
</Layout>
```

> You can then use the `auth` object to get the user's session in your server-side code.

Reviewer 2: ```suggestion
You can then use the `auth` object to get the user's session data in your server-side code. The following example personalizes page content by displaying an authenticated user's name:
```
Again, we always prefer to have a little introductory sentence that explains what exactly the code is doing, in addition to it just being an example of an abstract concept!

```astro title="src/pages/index.astro"
---

> import { auth } from "@/auth"; // import your Better Auth instance

Reviewer 2: ```suggestion
import { auth } from "../......"; // import your Better Auth instance
```
same comment here. this should point to a file that you've instructed someone to create

 
const session = await auth.api.getSession({
	headers: Astro.request.headers,
});
---

<p>{session.user?.name}</p>
```

> And you can use the `auth` object to protect your routes as well.

Reviewer 2: ```suggestion
You can also use the `auth` object to protect your routes using middleware. The following example checks whether a user trying to access a logged-in dashbord route is authenticated, and redirects them to the home page if not.
```
A short description of what is happening in the code helps prep the reader to know what to look for!

```ts title="src/middleware.ts"

> import { auth } from "@/auth";

Reviewer 2: ```suggestion
import { auth } from "../.....";
```

Author: Assuming they've placed the `auth.ts` file in the lib folder, which is common, `../../../lib/auth` should work.
